src/NUnitFramework/framework/Constraints/EqualConstraint{T}.cs

```diff
@@ -0,0 +1,362 @@
+// Copyright (c) Charlie Poole, Rob Prouse and Contributors. MIT License - see LICENSE.txt
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using NUnit.Framework.Internal;
+
+namespace NUnit.Framework.Constraints
+{
+    /// <summary>
+    /// EqualConstraint is able to compare an actual value with the
+    /// expected value provided in its constructor. Two objects are
+    /// considered equal if both are null, or if both have the same
+    /// value. NUnit has special semantics for some object types.
+    /// </summary>
+    public class EqualConstraint<T> : Constraint
+        where T : IEquatable<T>
+    {
+        #region Static and Instance Fields
+
+        private readonly T _expected;
+
+        private Tolerance _tolerance = Tolerance.Default;
+        private Func<T, T, bool> _comparer = null!;
+        private readonly NUnitEqualityComparer _fallbackComparer = new();
+
+        #endregion
+
+        #region Constructor
+        /// <summary>
+        /// Initializes a new instance of the <see cref="EqualConstraint"/> class.
+        /// </summary>
+        /// <param name="expected">The expected value.</param>
+        public EqualConstraint(T expected)
+            : base(expected)
```

Comment:
Wouldn't we not loose all 'generic' benefits of not boxing values when passing it to the non-generic base class as `object`.

Reply:
To be honest, I can't recall why I added that. Potentially it may, but only for the things that directly reference base class members rather than overriding generic equivalents. 

From my past attempts I found one of the challenges with making generic constraints "work" was internal pieces relying heavily on either `Constraint` or `IConstraint` as a base. I liked here how you were able to inherit from `Constraint` here while still overriding all the important pieces to get generic typing at the `ApplyTo()` level.

I've begun wondering if a transition approach to getting generic constraints into the codebase could be something like this PR where we still inherit from `Constraint` but override as much as possible - or at least where important - to enable generic typing. That would at least allow us to have the constraints "chain" together at the assertion level for strong-typing there without having to do all the plumbing in one go. One way that could work is if we added an `IConstraint<T>` which defines strongly-typed public members to carry generic typing through successive chained constraints while still inheriting from the base `Constraint` class that all the plumbing uses.

All just ideas. I was pleasantly surprised to see how easily generic typing could start to apply down at the comparer level without duplicating codepaths there

Reply:
Looking at this again, I think I better understand your question and I can't recall why I added this. I can try removing

Reply:
It appears it's used by the base class for XML generation, but I'll push a change shortly that at least cleans up a few references within this class.